skip a second call to shouldUseRecordFunction for BackendSelect ops

bdhirsh opened the pull request:
Highlights:
- Update BackendSelect kernels to use the new `redispatch` API, which does not use RecordFunction
- Remove `callWithDispatchKey` from the Dispatcher API, since it is no longer used anywhere.

As I pointed out in a [comment](https://github.com/pytorch/pytorch/pull/49354#issuecomment-764735318) in the commit below in this stack, the API changes from my previous commit make it really easy to avoid a second call to `shouldUseRecordFunction` in the BackendSelect ops.

Context: Factory functions are currently all implemented using the `BackendSelect` key, which results in two trips through the dispatcher: one to get to BackendSelect, and one to get to the actual backend that it computes. My changes in the previous PR make it easier to distinguish between an initial dispatch and subsequent dispatches in the chain, which shouldn't have to do as much work (namely, profiling and calls to `shouldUseRecordFunction`).

Perf win is most prominent for `torch::empty`: P171378499

Stack from [ghstack](https://github.com/ezyang/ghstack):
* #51363 update autograd kernels to use redispatch
* #51474 codegen'd API for redispatching
* #50848 inlining tensor.device()
* **#50891 skip a second call to shouldUseRecordFunction for BackendSelect ops**
* #49354 Plumbing TLS keys through the dispatcher
* #50963 remove unnecessary BoxedKernelWrapper specialization now that ops are all c10-full

Differential Revision: [D25999514](https://our.internmc.facebook.com/intern/diff/D25999514)

bdhirsh committed:
skip a second call to shouldUseRecordFunction for BackendSelect ops

[ghstack-poisoned]

bdhirsh committed:
Update on "skip a second call to shouldUseRecordFunction for BackendSelect ops"

As I pointed out in a [comment](https://github.com/pytorch/pytorch/pull/49354#issuecomment-764735318) in the commit below in this stack, the API changes from my previous commit make it really easy to avoid a second call to `shouldUseRecordFunction` in the BackendSelect ops.

Context: Factory functions are currently all implemented using the `BackendSelect` key, which results in two trips through the dispatcher: one to get to BackendSelect, and one to get to the actual backend that it computes. My changes in the previous PR make it easier to distinguish between an initial dispatch and subsequent dispatches in the chain, which shouldn't have to do as much work (namely, profiling and calls to `shouldUseRecordFunction`).

Callgrind benchmark: https://www.internalfb.com/intern/paste/P164935184/
The numbers look surprisingly good, I want to double check that I ran my benchmark correctly


Differential Revision: [D25999514](https://our.internmc.facebook.com/intern/diff/D25999514)

[ghstack-poisoned]

bdhirsh committed:
Update on "skip a second call to shouldUseRecordFunction for BackendSelect ops"

As I pointed out in a [comment](https://github.com/pytorch/pytorch/pull/49354#issuecomment-764735318) in the commit below in this stack, the API changes from my previous commit make it really easy to avoid a second call to `shouldUseRecordFunction` in the BackendSelect ops.

Context: Factory functions are currently all implemented using the `BackendSelect` key, which results in two trips through the dispatcher: one to get to BackendSelect, and one to get to the actual backend that it computes. My changes in the previous PR make it easier to distinguish between an initial dispatch and subsequent dispatches in the chain, which shouldn't have to do as much work (namely, profiling and calls to `shouldUseRecordFunction`).

Callgrind benchmark: https://www.internalfb.com/intern/paste/P164935184/
The numbers look surprisingly good, I want to double check that I ran my benchmark correctly


Differential Revision: [D25999514](https://our.internmc.facebook.com/intern/diff/D25999514)

[ghstack-poisoned]

bdhirsh committed:
Update on "skip a second call to shouldUseRecordFunction for BackendSelect ops"

As I pointed out in a [comment](https://github.com/pytorch/pytorch/pull/49354#issuecomment-764735318) in the commit below in this stack, the API changes from my previous commit make it really easy to avoid a second call to `shouldUseRecordFunction` in the BackendSelect ops.

Context: Factory functions are currently all implemented using the `BackendSelect` key, which results in two trips through the dispatcher: one to get to BackendSelect, and one to get to the actual backend that it computes. My changes in the previous PR make it easier to distinguish between an initial dispatch and subsequent dispatches in the chain, which shouldn't have to do as much work (namely, profiling and calls to `shouldUseRecordFunction`).

Callgrind benchmark: https://www.internalfb.com/intern/paste/P164935184/
The numbers look surprisingly good, I want to double check that I ran my benchmark correctly


Differential Revision: [D25999514](https://our.internmc.facebook.com/intern/diff/D25999514)

[ghstack-poisoned]

bdhirsh committed:
Update on "skip a second call to shouldUseRecordFunction for BackendSelect ops"

EDIT: just realized that I need to make some changes to ensure that BackendSelect kernels still hit RecordFunction the first time. The win won't be as big as I thought initially, but we should still see something. And regardless of the perf win, I think that the change in this PR is a logical improvement, since it differentiates between a call and a redispatch.

As I pointed out in a [comment](https://github.com/pytorch/pytorch/pull/49354#issuecomment-764735318) in the commit below in this stack, the API changes from my previous commit make it really easy to avoid a second call to `shouldUseRecordFunction` in the BackendSelect ops.

Context: Factory functions are currently all implemented using the `BackendSelect` key, which results in two trips through the dispatcher: one to get to BackendSelect, and one to get to the actual backend that it computes. My changes in the previous PR make it easier to distinguish between an initial dispatch and subsequent dispatches in the chain, which shouldn't have to do as much work (namely, profiling and calls to `shouldUseRecordFunction`).

Callgrind benchmark: coming soon


Differential Revision: [D25999514](https://our.internmc.facebook.com/intern/diff/D25999514)

[ghstack-poisoned]

bhosmer reviewed (approved):
LGTM, but re your earlier comment about making sure BackendSelect calls it at least once, is there anything tricky to review more closely there?

bdhirsh commented:
> LGTM, but re your earlier comment about making sure BackendSelect calls it at least once, is there anything tricky to review more closely there?

I confirmed with @ilia-cher that we have tests that it should be fine to only call `RecordFunction` once through the chain of redispatches, and that we have profiler tests guarding failures there.

Actually, this test (faithfully) failed initially in my PR: `python test/test_autograd.py TestAutograd.test_memory_profiler`.
Because I didn't notice the logic that explicitly tried to not call RecordFunction in the first kernel - `shouldRecord()`. Before deleting that, RecordFunction wasn't getting called at all in my PR for BackendSelect kernels. Removing it ensured that RecordFunction is called exactly once, when we redispatch to the backend kernel.

bdhirsh committed:
Update on "skip a second call to shouldUseRecordFunction for BackendSelect ops"

As I pointed out in a [comment](https://github.com/pytorch/pytorch/pull/49354#issuecomment-764735318) in the commit below in this stack, the API changes from my previous commit make it really easy to avoid a second call to `shouldUseRecordFunction` in the BackendSelect ops.

Context: Factory functions are currently all implemented using the `BackendSelect` key, which results in two trips through the dispatcher: one to get to BackendSelect, and one to get to the actual backend that it computes. My changes in the previous PR make it easier to distinguish between an initial dispatch and subsequent dispatches in the chain, which shouldn't have to do as much work (namely, profiling and calls to `shouldUseRecordFunction`).

Callgrind benchmark: https://www.internalfb.com/intern/paste/P164935184/

I didn't expect the win to be so large, since we have logic in `Dispatcher.h` to explicitly skip the usage of RecordFunction for BackendSelect kernels. As it exists in master, that check doesn't run until **after** we call `shouldRunRecordFunction`, so the main speedup is the savings from avoiding a second call to that function.

Some explanation / bookkeeping of the wins from an `at::empty` benchmark, which has a total speedup around ~7.5%
* not calling `shouldRunRecordFunction` a second time saved ~3.25%
* `callWithDispatchKey` was too big to get inlined, but `callWithPrecomputeDispatchKeySet` is smaller and thus gets inlined - ~1.7% shaved off from the lack of call overhead
* A trail of other smaller things (for example, the call to `shouldRunRecordFunction` itself got a little less expensive for some reason).


Differential Revision: [D25999514](https://our.internmc.facebook.com/intern/diff/D25999514)

[ghstack-poisoned]

bdhirsh committed:
Update on "skip a second call to shouldUseRecordFunction for BackendSelect ops"

As I pointed out in a [comment](https://github.com/pytorch/pytorch/pull/49354#issuecomment-764735318) in the commit below in this stack, the API changes from my previous commit make it really easy to avoid a second call to `shouldUseRecordFunction` in the BackendSelect ops.

Context: Factory functions are currently all implemented using the `BackendSelect` key, which results in two trips through the dispatcher: one to get to BackendSelect, and one to get to the actual backend that it computes. My changes in the previous PR make it easier to distinguish between an initial dispatch and subsequent dispatches in the chain, which shouldn't have to do as much work (namely, profiling and calls to `shouldUseRecordFunction`).

Callgrind benchmark: https://www.internalfb.com/intern/paste/P164935184/

I didn't expect the win to be so large, since we have logic in `Dispatcher.h` to explicitly skip the usage of RecordFunction for BackendSelect kernels. As it exists in master, that check doesn't run until **after** we call `shouldRunRecordFunction`, so the main speedup is the savings from avoiding a second call to that function.

Some explanation / bookkeeping of the wins from an `at::empty` benchmark, which has a total speedup around ~7.5%
* not calling `shouldRunRecordFunction` a second time saved ~3.25%
* `callWithDispatchKey` was too big to get inlined, but `callWithPrecomputeDispatchKeySet` is smaller and thus gets inlined - ~1.7% shaved off from the lack of call overhead
* A trail of other smaller things (for example, the call to `shouldRunRecordFunction` itself got a little less expensive for some reason).


Differential Revision: [D25999514](https://our.internmc.facebook.com/intern/diff/D25999514)

[ghstack-poisoned]

bdhirsh committed:
Update on "skip a second call to shouldUseRecordFunction for BackendSelect ops"

As I pointed out in a [comment](https://github.com/pytorch/pytorch/pull/49354#issuecomment-764735318) in the commit below in this stack, the API changes from my previous commit make it really easy to avoid a second call to `shouldUseRecordFunction` in the BackendSelect ops.

Context: Factory functions are currently all implemented using the `BackendSelect` key, which results in two trips through the dispatcher: one to get to BackendSelect, and one to get to the actual backend that it computes. My changes in the previous PR make it easier to distinguish between an initial dispatch and subsequent dispatches in the chain, which shouldn't have to do as much work (namely, profiling and calls to `shouldUseRecordFunction`).

Callgrind benchmark: https://www.internalfb.com/intern/paste/P164935184/

I didn't expect the win to be so large, since we have logic in `Dispatcher.h` to explicitly skip the usage of RecordFunction for BackendSelect kernels. As it exists in master, that check doesn't run until **after** we call `shouldRunRecordFunction`, so the main speedup is the savings from avoiding a second call to that function.

Some explanation / bookkeeping of the wins from an `at::empty` benchmark, which has a total speedup around ~7.5%
* not calling `shouldRunRecordFunction` a second time saved ~3.25%
* `callWithDispatchKey` was too big to get inlined, but `callWithPrecomputeDispatchKeySet` is smaller and thus gets inlined - ~1.7% shaved off from the lack of call overhead
* A trail of other smaller things (for example, the call to `shouldRunRecordFunction` itself got a little less expensive for some reason).


Differential Revision: [D25999514](https://our.internmc.facebook.com/intern/diff/D25999514)

[ghstack-poisoned]

bdhirsh committed:
Update on "skip a second call to shouldUseRecordFunction for BackendSelect ops"

As I pointed out in a [comment](https://github.com/pytorch/pytorch/pull/49354#issuecomment-764735318) in the commit below in this stack, the API changes from my previous commit make it really easy to avoid a second call to `shouldUseRecordFunction` in the BackendSelect ops.

Context: Factory functions are currently all implemented using the `BackendSelect` key, which results in two trips through the dispatcher: one to get to BackendSelect, and one to get to the actual backend that it computes. My changes in the previous PR make it easier to distinguish between an initial dispatch and subsequent dispatches in the chain, which shouldn't have to do as much work (namely, profiling and calls to `shouldUseRecordFunction`).

Callgrind benchmark: https://www.internalfb.com/intern/paste/P164935184/

I didn't expect the win to be so large, since we have logic in `Dispatcher.h` to explicitly skip the usage of RecordFunction for BackendSelect kernels. As it exists in master, that check doesn't run until **after** we call `shouldRunRecordFunction`, so the main speedup is the savings from avoiding a second call to that function.

Some explanation / bookkeeping of the wins from an `at::empty` benchmark, which has a total speedup around ~7.5%
* not calling `shouldRunRecordFunction` a second time saved ~3.25%
* `callWithDispatchKey` was too big to get inlined, but `callWithPrecomputeDispatchKeySet` is smaller and thus gets inlined - ~1.7% shaved off from the lack of call overhead
* A trail of other smaller things (for example, the call to `shouldRunRecordFunction` itself got a little less expensive for some reason).


Differential Revision: [D25999514](https://our.internmc.facebook.com/intern/diff/D25999514)

[ghstack-poisoned]

bdhirsh committed:
Update on "skip a second call to shouldUseRecordFunction for BackendSelect ops"

As I pointed out in a [comment](https://github.com/pytorch/pytorch/pull/49354#issuecomment-764735318) in the commit below in this stack, the API changes from my previous commit make it really easy to avoid a second call to `shouldUseRecordFunction` in the BackendSelect ops.

Context: Factory functions are currently all implemented using the `BackendSelect` key, which results in two trips through the dispatcher: one to get to BackendSelect, and one to get to the actual backend that it computes. My changes in the previous PR make it easier to distinguish between an initial dispatch and subsequent dispatches in the chain, which shouldn't have to do as much work (namely, profiling and calls to `shouldUseRecordFunction`).

Callgrind benchmark: https://www.internalfb.com/intern/paste/P164935184/

I didn't expect the win to be so large, since we have logic in `Dispatcher.h` to explicitly skip the usage of RecordFunction for BackendSelect kernels. As it exists in master, that check doesn't run until **after** we call `shouldRunRecordFunction`, so the main speedup is the savings from avoiding a second call to that function.

Some explanation / bookkeeping of the wins from an `at::empty` benchmark, which has a total speedup around ~7.5%
* not calling `shouldRunRecordFunction` a second time saved ~3.25%
* `callWithDispatchKey` was too big to get inlined, but `callWithPrecomputeDispatchKeySet` is smaller and thus gets inlined - ~1.7% shaved off from the lack of call overhead
* A trail of other smaller things (for example, the call to `shouldRunRecordFunction` itself got a little less expensive for some reason).


Differential Revision: [D25999514](https://our.internmc.facebook.com/intern/diff/D25999514)

[ghstack-poisoned]

bdhirsh committed:
Update on "skip a second call to shouldUseRecordFunction for BackendSelect ops"

As I pointed out in a [comment](https://github.com/pytorch/pytorch/pull/49354#issuecomment-764735318) in the commit below in this stack, the API changes from my previous commit make it really easy to avoid a second call to `shouldUseRecordFunction` in the BackendSelect ops.

Context: Factory functions are currently all implemented using the `BackendSelect` key, which results in two trips through the dispatcher: one to get to BackendSelect, and one to get to the actual backend that it computes. My changes in the previous PR make it easier to distinguish between an initial dispatch and subsequent dispatches in the chain, which shouldn't have to do as much work (namely, profiling and calls to `shouldUseRecordFunction`).

Callgrind benchmark: https://www.internalfb.com/intern/paste/P164935184/

I didn't expect the win to be so large, since we have logic in `Dispatcher.h` to explicitly skip the usage of RecordFunction for BackendSelect kernels. As it exists in master, that check doesn't run until **after** we call `shouldRunRecordFunction`, so the main speedup is the savings from avoiding a second call to that function.

Some explanation / bookkeeping of the wins from an `at::empty` benchmark, which has a total speedup around ~7.5%
* not calling `shouldRunRecordFunction` a second time saved ~3.25%
* `callWithDispatchKey` was too big to get inlined, but `callWithPrecomputeDispatchKeySet` is smaller and thus gets inlined - ~1.7% shaved off from the lack of call overhead
* A trail of other smaller things (for example, the call to `shouldRunRecordFunction` itself got a little less expensive for some reason).


Differential Revision: [D25999514](https://our.internmc.facebook.com/intern/diff/D25999514)

[ghstack-poisoned]

bdhirsh committed:
Update on "skip a second call to shouldUseRecordFunction for BackendSelect ops"

As I pointed out in a [comment](https://github.com/pytorch/pytorch/pull/49354#issuecomment-764735318) in the commit below in this stack, the API changes from my previous commit make it really easy to avoid a second call to `shouldUseRecordFunction` in the BackendSelect ops.

Context: Factory functions are currently all implemented using the `BackendSelect` key, which results in two trips through the dispatcher: one to get to BackendSelect, and one to get to the actual backend that it computes. My changes in the previous PR make it easier to distinguish between an initial dispatch and subsequent dispatches in the chain, which shouldn't have to do as much work (namely, profiling and calls to `shouldUseRecordFunction`).

Callgrind benchmark: https://www.internalfb.com/intern/paste/P164935184/

I didn't expect the win to be so large, since we have logic in `Dispatcher.h` to explicitly skip the usage of RecordFunction for BackendSelect kernels. As it exists in master, that check doesn't run until **after** we call `shouldRunRecordFunction`, so the main speedup is the savings from avoiding a second call to that function.

Some explanation / bookkeeping of the wins from an `at::empty` benchmark, which has a total speedup around ~7.5%
* not calling `shouldRunRecordFunction` a second time saved ~3.25%
* `callWithDispatchKey` was too big to get inlined, but `callWithPrecomputeDispatchKeySet` is smaller and thus gets inlined - ~1.7% shaved off from the lack of call overhead
* A trail of other smaller things (for example, the call to `shouldRunRecordFunction` itself got a little less expensive for some reason).


Differential Revision: [D25999514](https://our.internmc.facebook.com/intern/diff/D25999514)

[ghstack-poisoned]

bdhirsh committed:
Update on "skip a second call to shouldUseRecordFunction for BackendSelect ops"

As I pointed out in a [comment](https://github.com/pytorch/pytorch/pull/49354#issuecomment-764735318) in the commit below in this stack, the API changes from my previous commit make it really easy to avoid a second call to `shouldUseRecordFunction` in the BackendSelect ops.

Context: Factory functions are currently all implemented using the `BackendSelect` key, which results in two trips through the dispatcher: one to get to BackendSelect, and one to get to the actual backend that it computes. My changes in the previous PR make it easier to distinguish between an initial dispatch and subsequent dispatches in the chain, which shouldn't have to do as much work (namely, profiling and calls to `shouldUseRecordFunction`).

Callgrind benchmark: https://www.internalfb.com/intern/paste/P164935184/

I didn't expect the win to be so large, since we have logic in `Dispatcher.h` to explicitly skip the usage of RecordFunction for BackendSelect kernels. As it exists in master, that check doesn't run until **after** we call `shouldRunRecordFunction`, so the main speedup is the savings from avoiding a second call to that function.

Some explanation / bookkeeping of the wins from an `at::empty` benchmark, which has a total speedup around ~7.5%
* not calling `shouldRunRecordFunction` a second time saved ~3.25%
* `callWithDispatchKey` was too big to get inlined, but `callWithPrecomputeDispatchKeySet` is smaller and thus gets inlined - ~1.7% shaved off from the lack of call overhead
* A trail of other smaller things (for example, the call to `shouldRunRecordFunction` itself got a little less expensive for some reason).


Differential Revision: [D25999514](https://our.internmc.facebook.com/intern/diff/D25999514)

[ghstack-poisoned]

bdhirsh committed:
Update on "skip a second call to shouldUseRecordFunction for BackendSelect ops"

As I pointed out in a [comment](https://github.com/pytorch/pytorch/pull/49354#issuecomment-764735318) in the commit below in this stack, the API changes from my previous commit make it really easy to avoid a second call to `shouldUseRecordFunction` in the BackendSelect ops.

Context: Factory functions are currently all implemented using the `BackendSelect` key, which results in two trips through the dispatcher: one to get to BackendSelect, and one to get to the actual backend that it computes. My changes in the previous PR make it easier to distinguish between an initial dispatch and subsequent dispatches in the chain, which shouldn't have to do as much work (namely, profiling and calls to `shouldUseRecordFunction`).

Callgrind benchmark: https://www.internalfb.com/intern/paste/P164935184/

I didn't expect the win to be so large, since we have logic in `Dispatcher.h` to explicitly skip the usage of RecordFunction for BackendSelect kernels. As it exists in master, that check doesn't run until **after** we call `shouldRunRecordFunction`, so the main speedup is the savings from avoiding a second call to that function.

Some explanation / bookkeeping of the wins from an `at::empty` benchmark, which has a total speedup around ~7.5%
* not calling `shouldRunRecordFunction` a second time saved ~3.25%
* `callWithDispatchKey` was too big to get inlined, but `callWithPrecomputeDispatchKeySet` is smaller and thus gets inlined - ~1.7% shaved off from the lack of call overhead
* A trail of other smaller things (for example, the call to `shouldRunRecordFunction` itself got a little less expensive for some reason).


Differential Revision: [D25999514](https://our.internmc.facebook.com/intern/diff/D25999514)

[ghstack-poisoned]

bdhirsh committed:
Update on "skip a second call to shouldUseRecordFunction for BackendSelect ops"

As I pointed out in a [comment](https://github.com/pytorch/pytorch/pull/49354#issuecomment-764735318) in the commit below in this stack, the API changes from my previous commit make it really easy to avoid a second call to `shouldUseRecordFunction` in the BackendSelect ops.

Context: Factory functions are currently all implemented using the `BackendSelect` key, which results in two trips through the dispatcher: one to get to BackendSelect, and one to get to the actual backend that it computes. My changes in the previous PR make it easier to distinguish between an initial dispatch and subsequent dispatches in the chain, which shouldn't have to do as much work (namely, profiling and calls to `shouldUseRecordFunction`).

Callgrind benchmark: https://www.internalfb.com/intern/paste/P164935184/

I didn't expect the win to be so large, since we have logic in `Dispatcher.h` to explicitly skip the usage of RecordFunction for BackendSelect kernels. As it exists in master, that check doesn't run until **after** we call `shouldRunRecordFunction`, so the main speedup is the savings from avoiding a second call to that function.

Some explanation / bookkeeping of the wins from an `at::empty` benchmark, which has a total speedup around ~7.5%
* not calling `shouldRunRecordFunction` a second time saved ~3.25%
* `callWithDispatchKey` was too big to get inlined, but `callWithPrecomputeDispatchKeySet` is smaller and thus gets inlined - ~1.7% shaved off from the lack of call overhead
* A trail of other smaller things (for example, the call to `shouldRunRecordFunction` itself got a little less expensive for some reason).


Differential Revision: [D25999514](https://our.internmc.facebook.com/intern/diff/D25999514)

[ghstack-poisoned]

bdhirsh committed:
Update on "skip a second call to shouldUseRecordFunction for BackendSelect ops"

As I pointed out in a [comment](https://github.com/pytorch/pytorch/pull/49354#issuecomment-764735318) in the commit below in this stack, the API changes from my previous commit make it really easy to avoid a second call to `shouldUseRecordFunction` in the BackendSelect ops.

Context: Factory functions are currently all implemented using the `BackendSelect` key, which results in two trips through the dispatcher: one to get to BackendSelect, and one to get to the actual backend that it computes. My changes in the previous PR make it easier to distinguish between an initial dispatch and subsequent dispatches in the chain, which shouldn't have to do as much work (namely, profiling and calls to `shouldUseRecordFunction`).

Callgrind benchmark: https://www.internalfb.com/intern/paste/P164935184/

I didn't expect the win to be so large, since we have logic in `Dispatcher.h` to explicitly skip the usage of RecordFunction for BackendSelect kernels. As it exists in master, that check doesn't run until **after** we call `shouldRunRecordFunction`, so the main speedup is the savings from avoiding a second call to that function.

Some explanation / bookkeeping of the wins from an `at::empty` benchmark, which has a total speedup around ~7.5%
* not calling `shouldRunRecordFunction` a second time saved ~3.25%
* `callWithDispatchKey` was too big to get inlined, but `callWithPrecomputeDispatchKeySet` is smaller and thus gets inlined - ~1.7% shaved off from the lack of call overhead
* A trail of other smaller things (for example, the call to `shouldRunRecordFunction` itself got a little less expensive for some reason).


Differential Revision: [D25999514](https://our.internmc.facebook.com/intern/diff/D25999514)

[ghstack-poisoned]

bdhirsh committed:
Update on "skip a second call to shouldUseRecordFunction for BackendSelect ops"

As I pointed out in a [comment](https://github.com/pytorch/pytorch/pull/49354#issuecomment-764735318) in the commit below in this stack, the API changes from my previous commit make it really easy to avoid a second call to `shouldUseRecordFunction` in the BackendSelect ops.

Context: Factory functions are currently all implemented using the `BackendSelect` key, which results in two trips through the dispatcher: one to get to BackendSelect, and one to get to the actual backend that it computes. My changes in the previous PR make it easier to distinguish between an initial dispatch and subsequent dispatches in the chain, which shouldn't have to do as much work (namely, profiling and calls to `shouldUseRecordFunction`).

Callgrind benchmark: https://www.internalfb.com/intern/paste/P164935184/

I didn't expect the win to be so large, since we have logic in `Dispatcher.h` to explicitly skip the usage of RecordFunction for BackendSelect kernels. As it exists in master, that check doesn't run until **after** we call `shouldRunRecordFunction`, so the main speedup is the savings from avoiding a second call to that function.

Some explanation / bookkeeping of the wins from an `at::empty` benchmark, which has a total speedup around ~7.5%
* not calling `shouldRunRecordFunction` a second time saved ~3.25%
* `callWithDispatchKey` was too big to get inlined, but `callWithPrecomputeDispatchKeySet` is smaller and thus gets inlined - ~1.7% shaved off from the lack of call overhead
* A trail of other smaller things (for example, the call to `shouldRunRecordFunction` itself got a little less expensive for some reason).


Differential Revision: [D25999514](https://our.internmc.facebook.com/intern/diff/D25999514)

[ghstack-poisoned]

bdhirsh committed:
Update on "skip a second call to shouldUseRecordFunction for BackendSelect ops"

As I pointed out in a [comment](https://github.com/pytorch/pytorch/pull/49354#issuecomment-764735318) in the commit below in this stack, the API changes from my previous commit make it really easy to avoid a second call to `shouldUseRecordFunction` in the BackendSelect ops.

Context: Factory functions are currently all implemented using the `BackendSelect` key, which results in two trips through the dispatcher: one to get to BackendSelect, and one to get to the actual backend that it computes. My changes in the previous PR make it easier to distinguish between an initial dispatch and subsequent dispatches in the chain, which shouldn't have to do as much work (namely, profiling and calls to `shouldUseRecordFunction`).

Callgrind benchmark: https://www.internalfb.com/intern/paste/P164935184/

I didn't expect the win to be so large, since we have logic in `Dispatcher.h` to explicitly skip the usage of RecordFunction for BackendSelect kernels. As it exists in master, that check doesn't run until **after** we call `shouldRunRecordFunction`, so the main speedup is the savings from avoiding a second call to that function.

Some explanation / bookkeeping of the wins from an `at::empty` benchmark, which has a total speedup around ~7.5%
* not calling `shouldRunRecordFunction` a second time saved ~3.25%
* `callWithDispatchKey` was too big to get inlined, but `callWithPrecomputeDispatchKeySet` is smaller and thus gets inlined - ~1.7% shaved off from the lack of call overhead
* A trail of other smaller things (for example, the call to `shouldRunRecordFunction` itself got a little less expensive for some reason).


Differential Revision: [D25999514](https://our.internmc.facebook.com/intern/diff/D25999514)

[ghstack-poisoned]

bdhirsh committed:
Update on "skip a second call to shouldUseRecordFunction for BackendSelect ops"

As I pointed out in a [comment](https://github.com/pytorch/pytorch/pull/49354#issuecomment-764735318) in the commit below in this stack, the API changes from my previous commit make it really easy to avoid a second call to `shouldUseRecordFunction` in the BackendSelect ops.

Context: Factory functions are currently all implemented using the `BackendSelect` key, which results in two trips through the dispatcher: one to get to BackendSelect, and one to get to the actual backend that it computes. My changes in the previous PR make it easier to distinguish between an initial dispatch and subsequent dispatches in the chain, which shouldn't have to do as much work (namely, profiling and calls to `shouldUseRecordFunction`).

Callgrind benchmark: https://www.internalfb.com/intern/paste/P164935184/

I didn't expect the win to be so large, since we have logic in `Dispatcher.h` to explicitly skip the usage of RecordFunction for BackendSelect kernels. As it exists in master, that check doesn't run until **after** we call `shouldRunRecordFunction`, so the main speedup is the savings from avoiding a second call to that function.

Some explanation / bookkeeping of the wins from an `at::empty` benchmark, which has a total speedup around ~7.5%
* not calling `shouldRunRecordFunction` a second time saved ~3.25%
* `callWithDispatchKey` was too big to get inlined, but `callWithPrecomputeDispatchKeySet` is smaller and thus gets inlined - ~1.7% shaved off from the lack of call overhead
* A trail of other smaller things (for example, the call to `shouldRunRecordFunction` itself got a little less expensive for some reason).


Differential Revision: [D25999514](https://our.internmc.facebook.com/intern/diff/D25999514)

[ghstack-poisoned]

bdhirsh committed:
Update on "skip a second call to shouldUseRecordFunction for BackendSelect ops"

As I pointed out in a [comment](https://github.com/pytorch/pytorch/pull/49354#issuecomment-764735318) in the commit below in this stack, the API changes from my previous commit make it really easy to avoid a second call to `shouldUseRecordFunction` in the BackendSelect ops.

Context: Factory functions are currently all implemented using the `BackendSelect` key, which results in two trips through the dispatcher: one to get to BackendSelect, and one to get to the actual backend that it computes. My changes in the previous PR make it easier to distinguish between an initial dispatch and subsequent dispatches in the chain, which shouldn't have to do as much work (namely, profiling and calls to `shouldUseRecordFunction`).

Callgrind benchmark: https://www.internalfb.com/intern/paste/P164935184/

I didn't expect the win to be so large, since we have logic in `Dispatcher.h` to explicitly skip the usage of RecordFunction for BackendSelect kernels. As it exists in master, that check doesn't run until **after** we call `shouldRunRecordFunction`, so the main speedup is the savings from avoiding a second call to that function.

Some explanation / bookkeeping of the wins from an `at::empty` benchmark, which has a total speedup around ~7.5%
* not calling `shouldRunRecordFunction` a second time saved ~3.25%
* `callWithDispatchKey` was too big to get inlined, but `callWithPrecomputeDispatchKeySet` is smaller and thus gets inlined - ~1.7% shaved off from the lack of call overhead
* A trail of other smaller things (for example, the call to `shouldRunRecordFunction` itself got a little less expensive for some reason).


Differential Revision: [D25999514](https://our.internmc.facebook.com/intern/diff/D25999514)

[ghstack-poisoned]

bdhirsh committed:
Update on "skip a second call to shouldUseRecordFunction for BackendSelect ops"

As I pointed out in a [comment](https://github.com/pytorch/pytorch/pull/49354#issuecomment-764735318) in the commit below in this stack, the API changes from my previous commit make it really easy to avoid a second call to `shouldUseRecordFunction` in the BackendSelect ops.

Context: Factory functions are currently all implemented using the `BackendSelect` key, which results in two trips through the dispatcher: one to get to BackendSelect, and one to get to the actual backend that it computes. My changes in the previous PR make it easier to distinguish between an initial dispatch and subsequent dispatches in the chain, which shouldn't have to do as much work (namely, profiling and calls to `shouldUseRecordFunction`).

Callgrind benchmark: https://www.internalfb.com/intern/paste/P164935184/

I didn't expect the win to be so large, since we have logic in `Dispatcher.h` to explicitly skip the usage of RecordFunction for BackendSelect kernels. As it exists in master, that check doesn't run until **after** we call `shouldRunRecordFunction`, so the main speedup is the savings from avoiding a second call to that function.

Some explanation / bookkeeping of the wins from an `at::empty` benchmark, which has a total speedup around ~7.5%
* not calling `shouldRunRecordFunction` a second time saved ~3.25%
* `callWithDispatchKey` was too big to get inlined, but `callWithPrecomputeDispatchKeySet` is smaller and thus gets inlined - ~1.7% shaved off from the lack of call overhead
* A trail of other smaller things (for example, the call to `shouldRunRecordFunction` itself got a little less expensive for some reason).


Differential Revision: [D25999514](https://our.internmc.facebook.com/intern/diff/D25999514)

[ghstack-poisoned]

bdhirsh committed:
Update on "skip a second call to shouldUseRecordFunction for BackendSelect ops"

As I pointed out in a [comment](https://github.com/pytorch/pytorch/pull/49354#issuecomment-764735318) in the commit below in this stack, the API changes from my previous commit make it really easy to avoid a second call to `shouldUseRecordFunction` in the BackendSelect ops.

Context: Factory functions are currently all implemented using the `BackendSelect` key, which results in two trips through the dispatcher: one to get to BackendSelect, and one to get to the actual backend that it computes. My changes in the previous PR make it easier to distinguish between an initial dispatch and subsequent dispatches in the chain, which shouldn't have to do as much work (namely, profiling and calls to `shouldUseRecordFunction`).

Callgrind benchmark: https://www.internalfb.com/intern/paste/P164935184/

I didn't expect the win to be so large, since we have logic in `Dispatcher.h` to explicitly skip the usage of RecordFunction for BackendSelect kernels. As it exists in master, that check doesn't run until **after** we call `shouldRunRecordFunction`, so the main speedup is the savings from avoiding a second call to that function.

Some explanation / bookkeeping of the wins from an `at::empty` benchmark, which has a total speedup around ~7.5%
* not calling `shouldRunRecordFunction` a second time saved ~3.25%
* `callWithDispatchKey` was too big to get inlined, but `callWithPrecomputeDispatchKeySet` is smaller and thus gets inlined - ~1.7% shaved off from the lack of call overhead
* A trail of other smaller things (for example, the call to `shouldRunRecordFunction` itself got a little less expensive for some reason).


Differential Revision: [D25999514](https://our.internmc.facebook.com/intern/diff/D25999514)

[ghstack-poisoned]

bdhirsh committed:
Update on "skip a second call to shouldUseRecordFunction for BackendSelect ops"

Highlights:
- Remove `callWithDispatchKey` API from the dispatcher, since it's no longer used anywhere
- Update BackendSelect kernels to use the new `redispatch` API, which does not use RecordFunction


As I pointed out in a [comment](https://github.com/pytorch/pytorch/pull/49354#issuecomment-764735318) in the commit below in this stack, the API changes from my previous commit make it really easy to avoid a second call to `shouldUseRecordFunction` in the BackendSelect ops.

Context: Factory functions are currently all implemented using the `BackendSelect` key, which results in two trips through the dispatcher: one to get to BackendSelect, and one to get to the actual backend that it computes. My changes in the previous PR make it easier to distinguish between an initial dispatch and subsequent dispatches in the chain, which shouldn't have to do as much work (namely, profiling and calls to `shouldUseRecordFunction`).

Callgrind benchmark: https://www.internalfb.com/intern/paste/P164935184/

I didn't expect the win to be so large, since we have logic in `Dispatcher.h` to explicitly skip the usage of RecordFunction for BackendSelect kernels. As it exists in master, that check doesn't run until **after** we call `shouldRunRecordFunction`, so the main speedup is the savings from avoiding a second call to that function.

Some explanation / bookkeeping of the wins from an `at::empty` benchmark, which has a total speedup around ~7.5%
* not calling `shouldRunRecordFunction` a second time saved ~3.25%
* `callWithDispatchKey` was too big to get inlined, but `callWithPrecomputeDispatchKeySet` is smaller and thus gets inlined - ~1.7% shaved off from the lack of call overhead
* A trail of other smaller things (for example, the call to `shouldRunRecordFunction` itself got a little less expensive for some reason).


Differential Revision: [D25999514](https://our.internmc.facebook.com/intern/diff/D25999514)

[ghstack-poisoned]

bdhirsh committed:
Update on "skip a second call to shouldUseRecordFunction for BackendSelect ops"

Highlights:
- Remove `callWithDispatchKey` API from the dispatcher, since it's no longer used anywhere
- Update BackendSelect kernels to use the new `redispatch` API, which does not use RecordFunction


As I pointed out in a [comment](https://github.com/pytorch/pytorch/pull/49354#issuecomment-764735318) in the commit below in this stack, the API changes from my previous commit make it really easy to avoid a second call to `shouldUseRecordFunction` in the BackendSelect ops.

Context: Factory functions are currently all implemented using the `BackendSelect` key, which results in two trips through the dispatcher: one to get to BackendSelect, and one to get to the actual backend that it computes. My changes in the previous PR make it easier to distinguish between an initial dispatch and subsequent dispatches in the chain, which shouldn't have to do as much work (namely, profiling and calls to `shouldUseRecordFunction`).

Callgrind benchmark: https://www.internalfb.com/intern/paste/P164935184/

I didn't expect the win to be so large, since we have logic in `Dispatcher.h` to explicitly skip the usage of RecordFunction for BackendSelect kernels. As it exists in master, that check doesn't run until **after** we call `shouldRunRecordFunction`, so the main speedup is the savings from avoiding a second call to that function.

Some explanation / bookkeeping of the wins from an `at::empty` benchmark, which has a total speedup around ~7.5%
* not calling `shouldRunRecordFunction` a second time saved ~3.25%
* `callWithDispatchKey` was too big to get inlined, but `callWithPrecomputeDispatchKeySet` is smaller and thus gets inlined - ~1.7% shaved off from the lack of call overhead
* A trail of other smaller things (for example, the call to `shouldRunRecordFunction` itself got a little less expensive for some reason).


Differential Revision: [D25999514](https://our.internmc.facebook.com/intern/diff/D25999514)

[ghstack-poisoned]

bdhirsh committed:
Update on "skip a second call to shouldUseRecordFunction for BackendSelect ops"

Highlights:
- Remove `callWithDispatchKey` API from the dispatcher, since it's no longer used anywhere
- Update BackendSelect kernels to use the new `redispatch` API, which does not use RecordFunction


As I pointed out in a [comment](https://github.com/pytorch/pytorch/pull/49354#issuecomment-764735318) in the commit below in this stack, the API changes from my previous commit make it really easy to avoid a second call to `shouldUseRecordFunction` in the BackendSelect ops.

Context: Factory functions are currently all implemented using the `BackendSelect` key, which results in two trips through the dispatcher: one to get to BackendSelect, and one to get to the actual backend that it computes. My changes in the previous PR make it easier to distinguish between an initial dispatch and subsequent dispatches in the chain, which shouldn't have to do as much work (namely, profiling and calls to `shouldUseRecordFunction`).

Callgrind benchmark: https://www.internalfb.com/intern/paste/P164935184/

I didn't expect the win to be so large, since we have logic in `Dispatcher.h` to explicitly skip the usage of RecordFunction for BackendSelect kernels. As it exists in master, that check doesn't run until **after** we call `shouldRunRecordFunction`, so the main speedup is the savings from avoiding a second call to that function.

Some explanation / bookkeeping of the wins from an `at::empty` benchmark, which has a total speedup around ~7.5%
* not calling `shouldRunRecordFunction` a second time saved ~3.25%
* `callWithDispatchKey` was too big to get inlined, but `callWithPrecomputeDispatchKeySet` is smaller and thus gets inlined - ~1.7% shaved off from the lack of call overhead
* A trail of other smaller things (for example, the call to `shouldRunRecordFunction` itself got a little less expensive for some reason).


Differential Revision: [D25999514](https://our.internmc.facebook.com/intern/diff/D25999514)

[ghstack-poisoned]

bdhirsh committed:
Update on "skip a second call to shouldUseRecordFunction for BackendSelect ops"

Highlights:
- Update BackendSelect kernels to use the new `redispatch` API, which does not use RecordFunction
- Remove `callWithDispatchKey` from the Dispatcher API, since it is no longer used anywhere.

As I pointed out in a [comment](https://github.com/pytorch/pytorch/pull/49354#issuecomment-764735318) in the commit below in this stack, the API changes from my previous commit make it really easy to avoid a second call to `shouldUseRecordFunction` in the BackendSelect ops.

Context: Factory functions are currently all implemented using the `BackendSelect` key, which results in two trips through the dispatcher: one to get to BackendSelect, and one to get to the actual backend that it computes. My changes in the previous PR make it easier to distinguish between an initial dispatch and subsequent dispatches in the chain, which shouldn't have to do as much work (namely, profiling and calls to `shouldUseRecordFunction`).

Perf win is most prominent for `torch::empty`: P171378499


Differential Revision: [D25999514](https://our.internmc.facebook.com/intern/diff/D25999514)

[ghstack-poisoned]

bdhirsh committed:
Update on "skip a second call to shouldUseRecordFunction for BackendSelect ops"

Highlights:
- Update BackendSelect kernels to use the new `redispatch` API, which does not use RecordFunction
- Remove `callWithDispatchKey` from the Dispatcher API, since it is no longer used anywhere.

As I pointed out in a [comment](https://github.com/pytorch/pytorch/pull/49354#issuecomment-764735318) in the commit below in this stack, the API changes from my previous commit make it really easy to avoid a second call to `shouldUseRecordFunction` in the BackendSelect ops.

Context: Factory functions are currently all implemented using the `BackendSelect` key, which results in two trips through the dispatcher: one to get to BackendSelect, and one to get to the actual backend that it computes. My changes in the previous PR make it easier to distinguish between an initial dispatch and subsequent dispatches in the chain, which shouldn't have to do as much work (namely, profiling and calls to `shouldUseRecordFunction`).

Perf win is most prominent for `torch::empty`: P171378499


Differential Revision: [D25999514](https://our.internmc.facebook.com/intern/diff/D25999514)

[ghstack-poisoned]

bdhirsh committed:
Update on "skip a second call to shouldUseRecordFunction for BackendSelect ops"

Highlights:
- Update BackendSelect kernels to use the new `redispatch` API, which does not use RecordFunction
- Remove `callWithDispatchKey` from the Dispatcher API, since it is no longer used anywhere.

As I pointed out in a [comment](https://github.com/pytorch/pytorch/pull/49354#issuecomment-764735318) in the commit below in this stack, the API changes from my previous commit make it really easy to avoid a second call to `shouldUseRecordFunction` in the BackendSelect ops.

Context: Factory functions are currently all implemented using the `BackendSelect` key, which results in two trips through the dispatcher: one to get to BackendSelect, and one to get to the actual backend that it computes. My changes in the previous PR make it easier to distinguish between an initial dispatch and subsequent dispatches in the chain, which shouldn't have to do as much work (namely, profiling and calls to `shouldUseRecordFunction`).

Perf win is most prominent for `torch::empty`: P171378499


Differential Revision: [D25999514](https://our.internmc.facebook.com/intern/diff/D25999514)

[ghstack-poisoned]
